src/v/kafka/server/requests.cc

```diff
@@ -249,6 +249,14 @@ process_result_stages process_request(
   request_context&& ctx,
   ss::smp_service_group g,
   const session_resources& sres) {
+    auto& key = ctx.header().key;
+
+    if (
+      unlikely(ctx.sasl() && ctx.sasl()->complete() && ctx.sasl()->expired())
+      && key != sasl_handshake_handler::api::key
+      && key != sasl_authenticate_handler::api::key) {
+        throw sasl_session_expired_exception("Session expired");
```

Comment:
if the information is handy, it'd probably worth adding some context to this like how long the expiration was set to, and maybe the client id if isn't already logged at a higher level (or maybe higher level is where that extra context should go).